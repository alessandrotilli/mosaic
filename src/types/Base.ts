export interface IBase {
  dataCy?: string;
}

export interface ILoadable extends IBase {
  loading?: boolean;
}

// TODO: door open to allow explicit property localization
export interface ILocalizable extends IBase {
  localized?: boolean;
}
