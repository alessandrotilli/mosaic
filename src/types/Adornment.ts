import { IBase } from "./Base";

type IBadgeColor = "default" | "error" | "primary" | "secondary";

type IBadgeVariant = "dot" | "standard";

export interface IBadgeAdornment {
  color?: IBadgeColor;
  value: string;
  variant?: IBadgeVariant;
}

export interface IAdornment extends IBase {
  /**
   * Badge properties
   */
  badge?: IBadgeAdornment;
  /**
   * Text of tooltip
   */
  tooltip?: string;
}
