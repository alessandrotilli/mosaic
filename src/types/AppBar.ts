import { ILocalizable } from "./Base";
import { IBaseButton } from "./Button";
import { IBaseIconButton } from "./IconButton";

export interface IAppBar extends ILocalizable {
  actions?: IBaseIconButton[];
  menu?: IBaseIconButton;
  onTitleClick?: () => void;
  title?: string;
  userMenu?: IBaseButton[];
  username?: string;
}
