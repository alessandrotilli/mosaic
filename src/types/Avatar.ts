import { ILoadable } from "./Base";
import { Icons } from "./Icon";

export enum AvatarVariant {
  default = "circular",
  rounded = "rounded",
  squared = "square",
}

export interface IAvatar extends ILoadable {
  alt?: string;
  icon?: Icons;
  src?: string;
  text?: string;
  variant?: AvatarVariant;
}
