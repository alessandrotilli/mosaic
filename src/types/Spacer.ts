import { IBase } from "./Base";

export enum SpacerDirection {
  horizontal = "horizontal",
  vertical = "vertical",
}

export interface SpacerType extends IBase {
  direction?: SpacerDirection;
  level?: number;
}
