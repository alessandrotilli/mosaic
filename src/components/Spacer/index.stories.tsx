import React from "react";
import { select, number } from "@storybook/addon-knobs";
import { getDocumentationPage, StoriesWrapper } from "../../utils/stories";
import Spacer, { DATA_CY_DEFAULT } from ".";
import { SpacerDirection } from "../../types/Spacer";

export default {
  title: "Spacer",
  component: Spacer,
  parameters: {
    ...getDocumentationPage({
      basedOn: "@material-ui/core/Box",
      component: "Spacer",
      e2eTestInfo: {
        dataCyDefault: DATA_CY_DEFAULT,
      },
    }),
  },
};

export const Canvas = () => (
  <Spacer direction={select("direction", SpacerDirection, SpacerDirection.horizontal)} level={number("level", 1)} />
);

export const Horizontal = () => (
  <StoriesWrapper>
    <div style={{ backgroundColor: "#ff000f0a", display: "flex" }}>
      <span>Example</span>
      <span>Without</span>
    </div>
    <div style={{ backgroundColor: "#0f00ff0a", display: "flex" }}>
      <span>Example</span>
      <Spacer />
      <span>With</span>
    </div>
  </StoriesWrapper>
);

export const Vertical = () => (
  <StoriesWrapper>
    <div style={{ backgroundColor: "#ff000f0a", display: "flex", flexDirection: "column" }}>
      <span>Example</span>
      <span>Without</span>
    </div>
    <div style={{ backgroundColor: "#0f00ff0a", display: "flex", flexDirection: "column" }}>
      <span>Example</span>
      <Spacer direction={SpacerDirection.vertical} />
      <span>With</span>
    </div>
  </StoriesWrapper>
);
