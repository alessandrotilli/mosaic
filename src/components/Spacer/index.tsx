import React, { CSSProperties, FC, memo, useMemo } from "react";
import { useTheme } from "@mui/material";

import { ISpacer } from "../../types/Spacer";

export const DATA_CY_DEFAULT = "spacer";

const Spacer: FC<ISpacer> = ({ dataCy = DATA_CY_DEFAULT, direction = "horizontal", level: externalLevel = 1 }) => {
  const { spacing } = useTheme();

  const level = useMemo(() => (externalLevel < 0 || !externalLevel ? 1 : externalLevel), [externalLevel]);

  const height = useMemo(() => (direction === "horizontal" ? 0 : spacing(level)), [direction, level, spacing]);

  const width = useMemo(() => (direction === "horizontal" ? spacing(level) : 0), [direction, level, spacing]);

  const style = useMemo(
    (): CSSProperties => ({
      display: "block",
      height,
      width,
    }),
    [height, width]
  );

  return <div data-cy={dataCy} style={style} />;
};

export const MemoizedSpacer = memo(Spacer);

export default MemoizedSpacer;
