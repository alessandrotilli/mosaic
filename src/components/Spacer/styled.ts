import { Box as MUIBox, styled } from "@material-ui/core";
import { SpacerDirection } from "../../types/Spacer";

// TODO: this (props: any) here should be removed
export const StyledMUIBox = styled(MUIBox)((props: any) => {
  const { direction, level } = props;
  const size = level * 8;
  return {
    display: "block",
    height: direction === SpacerDirection.horizontal ? 0 : size,
    width: direction === SpacerDirection.horizontal ? size : 0,
  };
});
