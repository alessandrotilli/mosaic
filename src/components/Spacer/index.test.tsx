import React from "react";
import { mount, ReactWrapper } from "enzyme";
import { ISpacer, SpacerDirection } from "../../types/Spacer";
import Spacer, { DATA_CY_DEFAULT } from ".";

const defaultProps: ISpacer = {};

const getElement = (props?: ISpacer, dataCy = DATA_CY_DEFAULT): ReactWrapper => {
  const wrapper = mount(<Spacer {...defaultProps} {...props} />);
  return wrapper.find("div[data-cy='" + dataCy + "']");
};

describe("Spacer test suite:", () => {
  it("default", () => {
    const element = getElement();
    expect("default-props-check").toBeTruthy();
  });

  it("dataCy", () => {
    const element = getElement({ dataCy: "custom" }, "custom");
    expect(element).toBeInstanceOf(ReactWrapper);
    expect(element).toBeTruthy();
  });

  it("direction", () => {
    const element = getElement({ direction: SpacerDirection.vertical });
    expect(element.prop("direction")).toEqual(SpacerDirection.vertical);
  });

  it("level", () => {
    const element = getElement({ level: 2 });
    expect(element.prop("level")).toEqual(2);
  });
});
