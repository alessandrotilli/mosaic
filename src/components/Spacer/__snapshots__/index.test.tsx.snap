// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer test suite: dataCy 1`] = `
<div
  className="MuiBox-root MuiBox-root-9 Styled(MuiBox)-root-1 Styled(MuiBox)-root-8"
  data-cy="custom"
  direction="horizontal"
  level={1}
/>
`;

exports[`Spacer test suite: default 1`] = `
<div
  className="MuiBox-root MuiBox-root-5 Styled(MuiBox)-root-1 Styled(MuiBox)-root-4"
  data-cy="spacer"
  direction="horizontal"
  level={1}
/>
`;

exports[`Spacer test suite: direction 1`] = `
<div
  className="MuiBox-root MuiBox-root-13 Styled(MuiBox)-root-1 Styled(MuiBox)-root-12"
  data-cy="spacer"
  direction="vertical"
  level={1}
/>
`;

exports[`Spacer test suite: level 1`] = `
<div
  className="MuiBox-root MuiBox-root-17 Styled(MuiBox)-root-1 Styled(MuiBox)-root-16"
  data-cy="spacer"
  direction="horizontal"
  level={2}
/>
`;
